### ISSUE TITLE
Memory consumption exploded over the last few weeks

### ISSUE THREAD

**Issue by Toflar** [composer/composer#4928]
Hi everyone,

I know this is yet another issue related to memory consumption but I think there's really something strange going on here and I took notice of this phenomenon only two weeks ago or so.
Actually, it's very easy to reproduce and see the issue for yourself:

1.  `composer create-project symfony/framework-standard-edition`

Because this does an `install` behind the scenes, this executes quite normally. Then

2. `composer -vvv --profile update`

Here's my log file but as I said, you should be able to very easily reproduce:

[composer.log.txt](https://github.com/composer/composer/files/134062/composer.log.txt)

It literally explodes from 162 MB to 412 MB after reading the providers.
Can anybody reproduce and does anybody have an idea why this is?

**Comment by Seldaek** [comment]
That's the dependency resolution running and creating thousands of php objects which does use memory.. 400MB isn't particularly remarkable to be honest.

As it says in:

```
[412.1MB/25.24s] Analyzed 5952 packages to resolve dependencies
[412.1MB/25.24s] Analyzed 147562 rules to resolve dependencies
```

It created 150K rule objects, and had 6000 packages to look at.

**Comment by aschempp** [comment]
Composer just requires about 1GB in my latest installation. Strange is that the memory jumps by about 500MB when reading `symfony/security-guard`:

```
[237.6MB/66.95s] Writing /Users/aschempp/.composer/cache/repo/https---packagist.org/provider-symfony$property-info.json into cache
[237.3MB/66.99s] Downloading http://packagist.org/p/symfony/security-guard%24c16a50c9b371bfa40d5ce9dbf81441bd4a5212f59ac6ff20b0310b5d1b4ecec6.json
[237.9MB/67.09s] Writing /Users/aschempp/.composer/cache/repo/https---packagist.org/provider-symfony$security-guard.json into cache
[748.4MB/96.06s] Reading /Users/aschempp/.composer/cache/repo/https---packagist.org/provider-contao-legacy$dcawizard.json from cache
[748.4MB/96.08s] Reading /Users/aschempp/.composer/cache/repo/https---packagist.org/provider-contao-community-alliance$composer.json from cache
[753.3MB/96.26s] Reading /Users/aschempp/.composer/cache/repo/https---packagist.org/provider-contao-legacy$multicolumnwizard.json from cache
```

It then increases by another 300MB before starting to download:

```
[769.6MB/99.25s] Reading /Users/aschempp/.composer/cache/repo/https---packagist.org/provider-contao-legacy$notification-center.json from cache
[854.4MB/103.71s] Reading /Users/aschempp/.composer/cache/repo/https---packagist.org/provider-contao-legacy$namespaceclassloader.json from cache
[1038.1MB/123.63s] Analyzed 9889 packages to resolve dependencies
[1038.1MB/123.63s] Analyzed 441991 rules to resolve dependencies
```

Here's my `composer.json`.
```js
"require": {
    "php": ">=5.4.0",
    "symfony/symfony": "~2.7",
    "doctrine/orm": "^2.4.8",
    "symfony/swiftmailer-bundle": "~2.3",
    "symfony/monolog-bundle": "~2.7",
    "sensio/distribution-bundle": "^3.0.12",
    "sensio/framework-extra-bundle": "^3.0.2",
    "incenteev/composer-parameter-handler": "~2.0",
    "doctrine/annotations": "~1.2",

    "contao/core-bundle": "4.1.x-dev",
    "contao/installation-bundle": "1.0.x",
    "lexik/maintenance-bundle": "^2.0",

    "terminal42/contao-folderpage": "~2.0",
    "terminal42/notification_center": "~1.3",
    "terminal42/contao-leads": "~2.0",
    "terminal42/contao-ce-access": "~2.1",
    "menatwork/contao-multicolumnwizard": "~3.3",
    "terminal42/dcawizard": "~2.3",
    "codefog/contao-haste": "~4.12",

    "fxp/composer-asset-plugin": "~1.0",
    "bower-asset/fitvids": "dev-master",
    "bower-asset/underscore": "1.8.*",
    "bower-asset/intentionjs": "0.9.*",
    "bower-asset/add-to-homescreen": "~3.2",
    "bower-asset/jquery-selectric": "~1.9",
    "social-links/social-links": "^1.8.1",
    "happyr/google-api-bundle": "dev-feature/clients",
    "doctrine/doctrine-migrations-bundle": "1.0.x-dev",
    "doctrine/doctrine-cache-bundle": "^1.0@dev",
    "sixdays/opcache-bundle": "dev-master",
    "debril/rss-atom-bundle": "2.3.*",
    "raven/raven": "^0.13",
    "algolia/algolia-search-bundle": "^1.0",
    "phpoffice/phpexcel": "^1.8"
},
"require-dev": {
    "andres-montanez/magallanes": "dev-master",
    "bower-asset/susy": "2.2.*",
    "bower-asset/compass-mixins": "1.*",
    "phpunit/phpunit": "4.*"
},
```

**Comment by stof** [comment]
@aschempp the instantiation of rules is currently mixed with the downloading of provider files. This is why you see increases between downloads
